pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,554 @@
+const defaultBatchDeletionCount = 10
+
+func init() {
+	prometheus.MustRegister(outdatedVMIWorkloads)
+}
+
+type WorkloadUpdateController struct {
+	clientset             kubecli.KubevirtClient
+	queue                 workqueue.RateLimitingInterface
+	vmiInformer           cache.SharedIndexInformer
+	migrationInformer     cache.SharedIndexInformer
+	recorder              record.EventRecorder
+	migrationExpectations *controller.UIDTrackingControllerExpectations
+	kubeVirtInformer      cache.SharedIndexInformer
+	clusterConfig         *virtconfig.ClusterConfig
+	statusUpdater         *status.KVStatusUpdater
+
+	throttleIntervalSeconds int
+
+	// This lock protects cached data within this struct
+	// that is dynamic. The lock is held for the duration
+	// of the reconcile loop. The reconcile loop is already
+	// single threaded and only a single KubeVirt object
+	// may exist in the cluster at once. This lock is simply
+	// protection in the event that those assumptions ever
+	// change.
+	cacheLock sync.Mutex
+
+	// loop can become quite chatty during the update process. This optimization
+	// throttles how quickly the loop can fire since each loop execution is acting at
+	// a cluster wide level.
+	reconcileThrottleMap map[string]time.Time
+
+	lastDeletionBatch time.Time
+}
+
+type updateData struct {
+	allOutdatedVMIs        []*virtv1.VirtualMachineInstance
+	migratableOutdatedVMIs []*virtv1.VirtualMachineInstance
+	shutdownOutdatedVMIs   []*virtv1.VirtualMachineInstance
+
+	numActiveMigrations int
+}
+
+func NewWorkloadUpdateController(
+	vmiInformer cache.SharedIndexInformer,
+	migrationInformer cache.SharedIndexInformer,
+	kubeVirtInformer cache.SharedIndexInformer,
+	recorder record.EventRecorder,
+	clientset kubecli.KubevirtClient,
+	clusterConfig *virtconfig.ClusterConfig,
+) *WorkloadUpdateController {
+
+	c := &WorkloadUpdateController{
+		queue:                 workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		vmiInformer:           vmiInformer,
+		migrationInformer:     migrationInformer,
+		kubeVirtInformer:      kubeVirtInformer,
+		recorder:              recorder,
+		clientset:             clientset,
+		statusUpdater:         status.NewKubeVirtStatusUpdater(clientset),
+		migrationExpectations: controller.NewUIDTrackingControllerExpectations(controller.NewControllerExpectations()),
+		clusterConfig:         clusterConfig,
+
+		throttleIntervalSeconds: defaultThrottleIntervalSeconds,
+
+		reconcileThrottleMap: make(map[string]time.Time),
+	}
+
+	c.kubeVirtInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addKubeVirt,
+		DeleteFunc: c.deleteKubeVirt,
+		UpdateFunc: c.updateKubeVirt,
+	})
+
+	c.migrationInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMigration,
+		DeleteFunc: c.deleteMigration,
+		UpdateFunc: c.updateMigration,
+	})
+
+	return c
+}
+
+func (c *WorkloadUpdateController) getKubeVirtKey() (string, error) {
```

Comment:
Don't we have a common method for that yet? Maybe work having a common place for this?

Reply:
yup, i'll add that